packages/@dcl/ecs/src/systems/raycast.ts

```diff
@@ -0,0 +1,244 @@
+  ): void
+
+  /**
+   * @public
+   * Execute callback when the entity receives a RaycastResult component update.
+   * Uses a Vector3 global target position to calculate the ray direction
+   * @param entity - Entity to attach the callback
+   * @param callback - Function to execute when the entity's RaycastResult component is updated
+   * @param options - Raycast configuration options
+   */
+  registerGlobalTargetRaycast(
+    entity: Entity,
+    callback: RaycastSystemCallback,
+    options?: Partial<GlobalTargetRaycastOptions>
+  ): void
+
+  /**
+   * @public
+   * Execute callback when the entity receives a RaycastResult component update.
+   * Uses an target Entity value to calculate the ray direction
+   * @param entity - Entity to attach the callback
+   * @param callback - Function to execute when the entity's RaycastResult component is updated
+   * @param options - Raycast configuration options
+   */
+  registerTargetEntityRaycast(
+    entity: Entity,
+    callback: RaycastSystemCallback,
+    options?: Partial<TargetEntityRaycastOptions>
+  ): void
+}
+
+/**
+ * @internal
+ */
+export function createRaycastSystem(engine: IEngine): RaycastSystem {
+  const raycastComponent = components.Raycast(engine)
+  const raycastResultComponent = components.RaycastResult(engine)
+  const entitiesCallbackResultMap = new Map<Entity, { callback: RaycastSystemCallback }>()
+
+  const defaultOptions: RaycastSystemOptions = {
+    maxDistance: 16,
+    queryType: RaycastQueryType.RQT_HIT_FIRST,
+    continuous: false,
+    originOffset: { x: 0, y: 0, z: 0 },
+    collisionMask: ColliderLayer.CL_PHYSICS
+  }
+  const getLocalDirectionRaycastDefaultOptions = (
+    options: Partial<LocalDirectionRaycastOptions> = {}
+  ): RaycastSystemOptions => ({
+    ...defaultOptions,
+    directionRawValue: {
+      $case: 'localDirection',
+      localDirection: options.direction || { x: 0, y: 0, z: 1 }
+    },
+    ...options
+  })
+  const getGlobalDirectionRaycastDefaultOptions = (
+    options: Partial<GlobalDirectionRaycastOptions> = {}
+  ): RaycastSystemOptions => ({
+    ...defaultOptions,
+    directionRawValue: {
+      $case: 'globalDirection',
+      globalDirection: options.direction || { x: 0, y: 0, z: 1 }
+    },
+    ...options
+  })
+  const getGlobalTargetRaycastDefaultOptions = (
+    options: Partial<GlobalTargetRaycastOptions> = {}
+  ): RaycastSystemOptions => ({
+    ...defaultOptions,
+    directionRawValue: {
+      $case: 'globalTarget',
+      globalTarget: options.target || { x: 0, y: 0, z: 0 }
+    },
+    ...options
+  })
+  const getTargetEntityRaycastDefaultOptions = (
+    options: Partial<TargetEntityRaycastOptions> = {}
+  ): RaycastSystemOptions => ({
+    ...defaultOptions,
+    directionRawValue: {
+      $case: 'targetEntity',
+      targetEntity: options.targetEntity || 0
+    },
+    ...options
+  })
+
+  function registerRaycast(entity: Entity, callback: RaycastSystemCallback, options: RaycastSystemOptions) {
```

Comment:
you can only have one type of raycast declared right ? 
because we are always overriding it on each type of raycast

Reply:
Uhmm yes, for an entity only 1 type of raycast at a time (type -> direction type).
If the raycastSystem is called with different type of raycasts for 2 different entities, each entity will be configured with their respective raycast, and their callbacks stored for each entity respectively